### PR TITLE
Make network listener runnable cancellable

### LINKED ISSUE #11703: Intermittent crash in /html/semantics/embedded-content/the-iframe-element/change_parentage.html (Option::unwrap on a None value)
```
▶ CRASH [expected OK] /html/semantics/embedded-content/the-iframe-element/change_parentage.html
  │ 
  │ Xlib:  extension &#34;XFree86-VidModeExtension&#34; missing on display &#34;:0&#34;.
  │ ERROR:constellation::constellation: Panic: called `Option::unwrap()` on a `None` value
  │ ERROR:constellation::constellation: Backtrace:
  │ frame #0  - 0x00007f71d010466d - backtrace::backtrace::trace::hccde8df28b4db2a2
  │ frame #1  - 0x00007f71d01045f5 - backtrace::capture::Backtrace::new::h42f95930bb8c5ee8
  │ frame #2  - 0x00007f71cf15adc9 - _&lt;F as alloc..boxed..FnBox&lt;A&gt;&gt;::call_box::h89adae1a802be550
  │ frame #3  - 0x00007f71d00f5ea8 - util::panicking::initiate_panic_hook::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::he2b22674ad1748f3
  │ frame #4  - 0x00007f71d03c07cc - std::panicking::rust_panic_with_hook::h983af77c1a2e581b
  │ frame #5  - 0x00007f71d03dabd1 - std::panicking::begin_panic::he426e15a3766089a
  │ frame #6  - 0x00007f71d03c203a - std::panicking::begin_panic_fmt::hdddb415186c241e7
  │ frame #7  - 0x00007f71d03dab6e - rust_begin_unwind
  │ frame #8  - 0x00007f71d041116f - core::panicking::panic_fmt::hf4e16cb7f0d41a25
  │ frame #9  - 0x00007f71d0411448 - core::panicking::panic::h907815f47e914305
  │ frame #10 - 0x00007f71cef7cc81 - script::dom::event::Event::new_uninitialized::h05ae188a870e92b9
  │ frame #11 - 0x00007f71cef84b3b - script::dom::event::Event::new::h7a12c706eaf845bc
  │ frame #12 - 0x00007f71cefcb95c - script::dom::eventtarget::EventTarget::fire_event::h0d5b30039f5820d2
  │ frame #13 - 0x00007f71cf1cd7fa - script::task_source::dom_manipulation::DOMManipulationTask::handle_task::h0ea31f97e475cb7a
  │ frame #14 - 0x00007f71cf1652c5 - script::script_thread::ScriptThread::handle_msg_from_script::h47e979ae1e7fb676
  │ frame #15 - 0x00007f71cf1b1d9f - script::script_thread::ScriptThread::handle_msgs::_$u7b$$u7b$closure$u7d$$u7d$::hb50cf7fd97943b65
  │ frame #16 - 0x00007f71cf19a0e9 - script::script_thread::ScriptThread::handle_msgs::h1e1abab71191c950
  │ frame #17 - 0x00007f71cf159447 - std::panicking::try::call::h78dcd5319adf08fa
  │ frame #18 - 0x00007f71d03e4dcb - __rust_try
  │ frame #19 - 0x00007f71d03e4d6e - __rust_maybe_catch_panic
  │ frame #20 - 0x00007f71cf15a72d - _&lt;F as alloc..boxed..FnBox&lt;A&gt;&gt;::call_box::hce043844655ea6a9
  │ frame #21 - 0x00007f71d03d8d94 - std::sys::thread::Thread::new::thread_start::h9c883b6d445ece46
  │ frame #22 - 0x00007f71cc242181 - start_thread
  │ frame #23 - 0x00007f71cbd5947c - __clone
  │ frame #24 - 0x0000000000000000 - &lt;unknown&gt;
```

### LINKED ISSUE #12048: called `Option::unwrap()` on a `None` value (Event::new_uninitialized)
called `Option::unwrap()` on a `None` value
### URL:
https://features.wearemel.com/ill-retire-when-i-m-dead-why-continuing-to-work-is-good-for-a-man-s-health-c3053052d136?gi=47d56ab34b96
### Backtrace:
```
frame #0  - 0x000000011086c6ce - backtrace::backtrace::trace::hccde8df28b4db2a2
frame #1  - 0x000000011086c651 - backtrace::capture::Backtrace::new::h42f95930bb8c5ee8
frame #2  - 0x000000010fb3d086 - _<F as alloc..boxed..FnBox<A>>::call_box::h6059122298f825d8
frame #3  - 0x00000001107e0947 - util::panicking::initiate_panic_hook::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h2c0e77c593c00c08
frame #4  - 0x0000000110a54844 - std::panicking::rust_panic_with_hook::h983af77c1a2e581b
frame #5  - 0x0000000110a6d786 - std::panicking::begin_panic::he426e15a3766089a
frame #6  - 0x0000000110a55e88 - std::panicking::begin_panic_fmt::hdddb415186c241e7
frame #7  - 0x0000000110a6d3df - rust_begin_unwind
frame #8  - 0x0000000110a94260 - core::panicking::panic_fmt::hf4e16cb7f0d41a25
frame #9  - 0x0000000110a9455c - core::panicking::panic::h907815f47e914305
frame #10 - 0x000000010f938895 - script::dom::event::Event::new_uninitialized::he3ccb94702598d61
frame #11 - 0x000000010f940aa7 - script::dom::event::Event::new::h115b505636eaa661
frame #12 - 0x000000010fa0d312 - script::dom::htmlscriptelement::HTMLScriptElement::dispatch_event::hfd4452ea712f8efd
frame #13 - 0x000000010f927241 - script::dom::htmlscriptelement::HTMLScriptElement::execute::hd6eb39d1b33b28da
frame #14 - 0x000000010f37580c - script::dom::document::Document::finish_load::h1ce7e49c25d85fc2
frame #15 - 0x000000010fa069b8 - _<script..dom..htmlscriptelement..ScriptContext as net_traits..AsyncResponseListener>::response_complete::h536287b80705375c
frame #16 - 0x000000010fa0c5c5 - _<script..network_listener..ListenerRunnable<A, Listener> as script..script_thread..Runnable>::handler::h213e3c65c1a0f2f0
frame #17 - 0x000000010fb4895c - script::script_thread::ScriptThread::handle_msg_from_script::hd1f5e7b2423a8c64
frame #18 - 0x000000010fba4a71 - script::script_thread::ScriptThread::handle_msgs::_$u7b$$u7b$closure$u7d$$u7d$::hc67d8b381c013b79
frame #19 - 0x000000010fb8a73f - script::script_thread::ScriptThread::handle_msgs::h236c5be60a207bea
frame #20 - 0x000000010fb3b907 - std::panicking::try::call::h645a328f039d97c6
frame #21 - 0x0000000110a7046b - __rust_try
frame #22 - 0x0000000110a70405 - __rust_maybe_catch_panic
frame #23 - 0x000000010fb3cab4 - _<F as alloc..boxed..FnBox<A>>::call_box::h3db95636eb747818
frame #24 - 0x0000000110a6bee8 - std::sys::thread::Thread::new::thread_start::h9c883b6d445ece46
frame #25 - 0x00007fff94f6f99c - _pthread_body
frame #26 - 0x00007fff94f6f919 - _pthread_start

```

### PR DESCRIPTION
**Pull request by cbrewster**
This keeps scripts from executing after its script thread has been shut down.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [x] These changes fix (maybe) #12048 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12277)

### CODE REVIEW

**Comment by jdm** [comment]
Here's a test that might work:
* parent.html adds an iframe loading child.html in a cross-origin manner (to ensure a separate script thread)
* child.html has a load event listener that appends an external script to the document
* this script uses wptserve's network filter to delay the script's network response by one second (`foo.js?pipe=trickle(d1)`)
* parent.html's iframe load event removes the iframe, then finishes the test three seconds later

**Comment by ConnorGBrewster** [comment]
@jdm turns out, the panic only occurs if the  iframe is same-origin, which seems bad since they should share the same script thread.

**Comment by ConnorGBrewster** [comment]
In fact this means that the test will still fail, although with these changes, it will no longer panic.

**Comment by ConnorGBrewster** [comment]
I added a test case that fails w/o the cancellable changes.

**Comment by ConnorGBrewster** [comment]
I want to run a few tries on this to make sure it fixes #11703.
@bors-servo try

**Comment by jdm** [comment]
If the test in question shuts down the shared script thread, that's a problem that we should be fixing :)

**Comment by ConnorGBrewster** [comment]
@jdm after the iframe is removed, the page still runs fine with the test results. So I assume that means the script thread for the parent isn't shut down.

**Comment by jdm** [comment]
True!

File: tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/script-not-executed-after-shutdown.js [hunk 1]
```diff
@@ -0,0 +1 @@
+// this script does nothing
```
Comment (jdm) [inline tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/script-not-executed-after-shutdown.js]: This script should call a function in the parent page which would make the test fail. Something like
`window.script_executed = t.unreached_func("script executed in removed iframe")` in the parent, and in the child: `script_executed = parent.script_executed;` then invoke `script_executed()` in this JS file.

**Comment by ConnorGBrewster** [comment]
Well this does not seem to fix the intermittent :(

**Comment by jdm** [comment]
The backtrace of the most recent panic is coming from a DOMManipulationTask runnable, not a network one.

**Comment by ConnorGBrewster** [comment]
@jdm is it possible to do something similar with that runnable?

**Comment by jdm** [comment]
What I'm wondering is why we're still running events for a script thread that's been shut down...

**Comment by jdm** [comment]
Oh, it's not the script thread that is shut down, it's a window that's had its JS context information cleared because it is not supposed to be needed any more. Bah.

**Comment by jdm** [comment]
There are a bunch of things in DOMManipulationTask that use Runnable which could theoretically be hooked the same way. The FireEvent and FireSimpleEvent messages would need special handling to ignore those messages if they're targeted at elements of a window that are closed, I guess ;(

**Comment by Ms2ger** [comment]
Fwiw, it's never been clear to me what the JSContext clearing is for.

**Comment by jdm** [comment]
The goal was to avoid hanging on to JS memory longer than necessary, since clearing it can end up removing the last user of the JS runtime.

**Comment by ConnorGBrewster** [comment]
I think I will leave this one as it is, and open a new PR when trying to solve the issue with `DOMManipulationTask`.

**Comment by Ms2ger** [comment]
I don't think so; the script thread should keep hanging on to the runtime/context.

**Comment by Manishearth** [comment]
So the wrapper uses Relaxed atomic ordering. Which means that even if the bool is written to from the script thread, it's not necessary that the other thread will see this write in time.

We should either use SeqCst ordering, or Release on the script thread side writes and acquire from the wrapper reads.


Everything else :lgtm: 

---

Reviewed 5 of 6 files at r1, 1 of 5 files at r3, 1 of 1 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/script-not-executed-after-shutdown-child.html, line 9 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12277#-KM9wlB2_edZo_qsW_qa:-KM9wlB2_edZo_qsW_qb:278106375) ([raw file](https://github.com/servo/servo/blob/411acaf8d86266cc79196172e0d02dea0fa569fe/tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/script-not-executed-after-shutdown-child.html#L9)):*
> ```HTML
>   s = document.createElement('script');
>   s.type = 'text/javascript';
>   s.src = 'script-not-executed-after-shutdown.js?pipe=trickle(d1)';
> ```

Perhaps a longer delay would ensure intermittents won't happen on a system experiencing heavy load (like our testing infrastructure at times)?

(Bump the 3000 up as well, of course)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12277#-:-KM9wBADjO5G-GHigNcr:-1594332496)*

**Comment by ConnorGBrewster** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/script-not-executed-after-shutdown-child.html, line 9 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12277#-KM9wlB2_edZo_qsW_qa:-KMA11SzhykGtGlxjpsd:-499146009) ([raw file](https://github.com/servo/servo/blob/411acaf8d86266cc79196172e0d02dea0fa569fe/tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/script-not-executed-after-shutdown-child.html#L9)):*
<details><summary><i>Previously, Manishearth (Manish Goregaokar) wrote…</i></summary>
> Perhaps a longer delay would ensure intermittents won't happen on a system experiencing heavy load (like our testing infrastructure at times)?
> 
> (Bump the 3000 up as well, of course)
</details>
Done.

---

*[tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/script-not-executed-after-shutdown.js, line 1 \[r4\]](https://reviewable.io:443/reviews/servo/servo/12277#-KLyEoThVOl6OBnjLkCb-r4-1:-KMA117GbgEEHubOHHuE:-499146009) ([raw file](https://github.com/servo/servo/blob/0cd7688180971ba7bd119c28bd5333f582dd93d9/tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/script-not-executed-after-shutdown.js#L1)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> This script should call a function in the parent page which would make the test fail. Something like
> 
> `window.script_executed = t.unreached_func("script executed in removed iframe")` in the parent, and in the child: `script_executed = parent.script_executed;` then invoke `script_executed()` in this JS file.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12277)*

**Comment by Manishearth** [comment]
@bors-servo r+

---

Reviewed 6 of 6 files at r6, 4 of 4 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12277#-:-KMADdpaN2QaH1VKt5aE:1236647940)*

**Comment by Manishearth** [comment]
@bors-servo retry
 - Seemed to fail compile without any error? Wat.

**Pull request merged**